os/drivers/input/touch.c

```diff
@@ -0,0 +1,375 @@
+
+static inline int touch_semtake(FAR sem_t *sem, bool errout)
+{
+	/* Take a count from the semaphore, possibly waiting */
+	if (sem_wait(sem) < 0) {
+		/* EINTR is the only error that we expect */
+		int errcode = get_errno();
+		DEBUGASSERT(errcode == EINTR);
+		if (errout) {
+			return errcode;
+		}
+	}
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_semgive
+ ****************************************************************************/
+
+static inline void touch_semgive(sem_t *sem)
+{
+	sem_post(sem);
+}
+
+/****************************************************************************
+ * Name: touch_open
+ ****************************************************************************/
+
+static int touch_open(FAR struct file *filep)
+{
+	FAR struct touch_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	if (priv->crefs == 0) {
+		priv->dev->touch_enable(priv->dev);
+	}
+	priv->crefs++;
+	DEBUGASSERT(priv->crefs > 0);
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_close
+ ****************************************************************************/
+
+static int touch_close(FAR struct file *filep)
+{
+	FAR struct touch_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	DEBUGASSERT(priv->crefs > 0);
+	priv->crefs--;
+	if (priv->crefs == 0) {
+		priv->dev->touch_disable(priv->dev);
+#ifndef CONFIG_DISABLE_POLL
+		/* Check if this file is registered in a list of waiters for polling.
+		* For example, when task A is blocked by calling poll and task B try to terminate task A,
+		* a pollfd of A remains in this list. If it is, it should be cleared.
+		*/
+		(void)touch_semtake(&priv->pollsem, false);
+		for (int i = 0; i < CONFIG_TOUCH_NPOLLWAITERS; i++) {
+			struct pollfd *fds = priv->fds[i];
+			if (fds && (FAR struct file *)fds->filep == filep) {
+				priv->fds[i] = NULL;
+			}
+		}
+		touch_semgive(&priv->pollsem);
+#endif
+	}
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_read
+ ****************************************************************************/
+
+static ssize_t touch_read(FAR struct file *filep, FAR char *buffer, size_t buflen)
+{
+	FAR struct touch_s *priv;
+	size_t outlen;
+	int ret;
+
+	priv = filep->f_inode->i_private;
+
+	if (!priv || !buffer || buflen < 1) {
+		return -EINVAL;
+	}
+	/* Wait for semaphore to prevent concurrent reads */
+	touch_semtake(&priv->sem, false);
+
+	/* Read the touch data, only if screen has been touched or if we're waiting for touch up */
+	outlen = sizeof(struct touch_sample_s);
+	ret = priv->dev->touch_read(priv->dev, buffer);
```

Comment:
You should use ioctl instead of read & write.
If another device should use some specific structure to bring data, then write & read will not bring data.
That's why all character device use ioctl instead of read / write